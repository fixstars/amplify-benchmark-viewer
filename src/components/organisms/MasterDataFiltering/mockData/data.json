[
  { "name": "Tsp" },
  { "name": "Qap" },
  { "name": "MaxCut" },
  { "name": "Sudoku" },
  { "name": "Cvrp" }
]
