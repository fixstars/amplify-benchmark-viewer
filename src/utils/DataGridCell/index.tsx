/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { LinkButton } from 'components/atoms/LinkButton'

const clientColors = [
  '#0f66dc',
  '#f57bae',
  '#e75e75',
  '#e5f97c',
  '#1b000d',
  '#f6a695',
  '#692981',
  '#3cde29',
  '#591414',
  '#d66594',
  '#2d884a',
  '#400763',
  '#9d19f4',
  '#fcdeb9',
  '#1f4c48',
  '#909743',
  '#2dc1c8',
  '#c17913',
  '#b2a1f6',
]

const instanceColors = [
  '#ff66ff',
  '#0f66dc',
  '#f57bae',
  '#e75e75',
  '#e5f97c',
  '#1b000d',
  '#f6a695',
  '#692981',
  '#3cde29',
  '#591414',
  '#d66594',
  '#2d884a',
  '#400763',
  '#9d19f4',
  '#fcdeb9',
  '#718de6',
  '#1f4c48',
  '#909743',
  '#2dc1c8',
  '#c17913',
  '#b2a1f6',
  '#9a916d',
  '#406a0f',
  '#004b31',
  '#18729f',
  '#9997c3',
  '#9cf0ec',
  '#ce843c',
  '#2d326e',
  '#76ed60',
  '#d7f7e8',
  '#bb15d7',
  '#87a7d8',
  '#94053a',
  '#e5c8a6',
  '#e2f2c0',
]

export interface ProblemInstances {
  readonly [key: string]: ReadonlyArray<string> | undefined
}

const renderClientsCell = (key: string, clients: ReadonlyArray<string>) => {
  const getBackgroundColor = (client: string): string => {
    const cache = JSON.parse(localStorage.getItem('clientColorCache') ?? '{}')
    let cacheIndex = JSON.parse(
      localStorage.getItem('clientColorIndexCache') ?? '-1',
    )
    let backgroundColor = cache[client]

    if (backgroundColor === undefined) {
      const length = clientColors.length
      cacheIndex = cacheIndex >= length ? 0 : cacheIndex + 1
      backgroundColor = clientColors[cacheIndex]
      cache[client] = backgroundColor
      localStorage.setItem('clientColorCache', JSON.stringify(cache))
      localStorage.setItem('clientColorIndexCache', JSON.stringify(cacheIndex))
      return backgroundColor
    }

    return backgroundColor
  }

  return clients.map((client) => (
    <div key={`${key}_${client}`} style={{ marginRight: 10 }}>
      <LinkButton
        label={client}
        link={`/clients/${client}`}
        backgroundColor={getBackgroundColor(client)}
      />
    </div>
  ))
}

const renderProblemInstancesCell = (
  key: string,
  problemInstances: ProblemInstances,
) => {
  const getBackgroundColor = (instance: string): string => {
    const cache = JSON.parse(localStorage.getItem('instanceColorCache') ?? '{}')
    let cacheIndex = JSON.parse(
      localStorage.getItem('instanceColorIndexCache') ?? '-1',
    )
    let backgroundColor = cache[instance]

    if (backgroundColor === undefined) {
      const length = instanceColors.length
      cacheIndex = cacheIndex >= length ? 0 : cacheIndex + 1
      backgroundColor = instanceColors[cacheIndex]
      cache[instance] = backgroundColor
      localStorage.setItem('instanceColorCache', JSON.stringify(cache))
      localStorage.setItem(
        'instanceColorIndexCache',
        JSON.stringify(cacheIndex),
      )
      return backgroundColor
    }

    return backgroundColor
  }
  const pairs = Object.entries(problemInstances)
  pairs.sort(function (p1, p2) {
    const p1Key = p1[0]
    const p2Key = p2[0]
    if (p1Key < p2Key) {
      return -1
    }
    if (p1Key > p2Key) {
      return 1
    }
    return 0
  })
  problemInstances = Object.fromEntries(pairs)
  return Object.keys(problemInstances).map((className) => (
    <div
      key={`${key}_${className}`}
      style={{ margin: '5px 0 5px 0', display: 'flex' }}
    >
      <span style={{ marginRight: 10 }}>{className}:</span>
      <div style={{ display: 'flex', flexWrap: 'wrap' }}>
        {problemInstances[className]?.map((instance) => (
          <div
            key={`${key}_${className}_${instance}`}
            style={{ marginRight: 10, marginBottom: 10 }}
          >
            <LinkButton
              label={instance}
              link={`/classes/${className}/instances/${instance}`}
              backgroundColor={getBackgroundColor(instance)}
            />
          </div>
        ))}
      </div>
    </div>
  ))
}

export { renderClientsCell, renderProblemInstancesCell }
