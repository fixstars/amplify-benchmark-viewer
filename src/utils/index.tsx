export * from './DataGridCell'
export * from './statistics'
