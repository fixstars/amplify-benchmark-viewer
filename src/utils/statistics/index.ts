function quartile(data: readonly number[], q: number) {
  const sortedData = data.slice().sort((a, b) => a - b)
  const pos = Math.floor((sortedData.length - 1) * q)
  return sortedData[pos]
}
function quartile0(data: readonly number[]) {
  return quartile(data, 0)
}
function quartile25(data: readonly number[]) {
  return quartile(data, 0.25)
}

function quartile50(data: readonly number[]) {
  return quartile(data, 0.5)
}

function quartile75(data: readonly number[]) {
  return quartile(data, 0.75)
}

function quartile100(data: readonly number[]) {
  return quartile(data, 1)
}

export { quartile0, quartile25, quartile50, quartile75, quartile100 }
