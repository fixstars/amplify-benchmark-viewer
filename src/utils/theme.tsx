/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { createTheme } from '@mui/material/styles'

export const theme = createTheme({
  palette: {
    primary: {
      main: '#111',
    },
  },
  typography: {
    h2: {
      fontSize: '1.5rem',
    },
  },
  components: {
    MuiTooltip: {
      styleOverrides: {
        tooltip: {
          backgroundColor: '#111',
          color: '#f4f1f4',
          font: '10px',
        },
      },
    },
  },
})
