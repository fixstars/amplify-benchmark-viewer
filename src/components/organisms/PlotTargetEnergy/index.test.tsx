import { render } from '@testing-library/react'
import { mockPlot } from 'utils/test'
import mockData from 'utils/test/mocks/data/mockReportData.json'

import { PlotTargetEnergy } from '.'

describe('<PlotTargetEnergy />', () => {
  it('Rendered well', async () => {
    const { container } = render(
      <PlotTargetEnergy
        data={mockData}
        instance="r_instance"
        useHistory
        label=""
      />,
    )

    const plot = mockPlot.mock.calls[0][0]
    expect(plot.style).toEqual({ flex: 1, height: '100%' })
    expect(plot.config).toEqual({
      toImageButtonOptions: {
        filename: 'TargetEnergy_r_instance_',
      },
      responsive: true,
    })
    expect(plot.layout).toEqual({
      showlegend: true,
      updatemenus: [
        {
          buttons: [
            {
              args: [
                {},
                {
                  xaxis: {
                    title: 'sampling_time[ms]',
                    type: 'log',
                  },
                },
              ],
              label: 'Log Scale',
              method: 'update',
            },
            {
              args: [
                {},
                {
                  xaxis: {
                    title: 'sampling_time[ms]',
                    type: 'linear',
                  },
                },
              ],
              label: 'Linear Scale',
              method: 'update',
            },
          ],
          xanchor: 'left',
          y: 1.2,
          yanchor: 'top',
        },
        {
          buttons: [
            {
              args: [
                {
                  visible: [true, true, true, true, true, true],
                },
              ],
              label: 'Scatter',
              method: 'update',
            },
            {
              args: [
                {
                  visible: [true, true, true, true, true, true],
                },
              ],
              label: 'Box',
              method: 'update',
            },
          ],
          x: 0.2,
          xanchor: 'left',
          y: 1.2,
          yanchor: 'top',
        },
      ],
      xaxis: {
        autorange: true,
        title: 'sampling_time[ms]',
        type: 'log',
      },
      yaxis: {
        autorange: true,
        title: 'target_energy',
        type: 'linear',
      },
    })
    expect(plot.data).toHaveLength(7)
    expect(plot.data).toMatchObject([
      {
        fill: 'toself',
        fillcolor: '#1f77b455',
        legendgroup: '1_a_client(v0.6.4-54-ge56964d-V100)',
        mode: 'lines',
        name: '1_a_client(v0.6.4-54-ge56964d-V100)',
        showlegend: false,
        text: 'max-min',
        type: 'scatter',
      },
      {
        fill: 'toself',
        fillcolor: '#1f77b466',
        legendgroup: '1_a_client(v0.6.4-54-ge56964d-V100)',
        mode: 'lines',
        name: '1_a_client(v0.6.4-54-ge56964d-V100)',
        showlegend: false,
        text: '75%-25%',
        type: 'scatter',
      },
      {
        fillcolor: '#1f77b4ff',
        legendgroup: '1_a_client(v0.6.4-54-ge56964d-V100)',
        mode: 'lines',
        name: '1_a_client(v0.6.4-54-ge56964d-V100)',
        showlegend: true,
        type: 'scatter',
      },
      {
        fill: 'toself',
        fillcolor: '#ff7f0e55',
        legendgroup: '2_a_client(v0.6.4-54-ge56964d-A100)',
        mode: 'lines',
        name: '2_a_client(v0.6.4-54-ge56964d-A100)',
        showlegend: false,
        text: 'max-min',
        type: 'scatter',
      },
      {
        fill: 'toself',
        fillcolor: '#ff7f0e66',
        legendgroup: '2_a_client(v0.6.4-54-ge56964d-A100)',
        mode: 'lines',
        name: '2_a_client(v0.6.4-54-ge56964d-A100)',
        showlegend: false,
        text: '75%-25%',
        type: 'scatter',
      },
      {
        fillcolor: '#ff7f0eff',
        legendgroup: '2_a_client(v0.6.4-54-ge56964d-A100)',
        mode: 'lines',
        name: '2_a_client(v0.6.4-54-ge56964d-A100)',
        showlegend: true,
        type: 'scatter',
      },
      {
        line: {
          color: 'black',
          dash: 'dot',
          width: 4,
        },
        mode: 'lines',
        name: 'best known',
        showlegend: false,
        x: [265.358596, 1000000],
        y: [59030, 59030],
      },
    ])
    expect(container).toMatchSnapshot()
  })

  it('Rendered well when label is changed', async () => {
    render(
      <PlotTargetEnergy
        data={mockData}
        instance="r_instance"
        useHistory
        label="20220927_173640"
      />,
    )

    const plot = mockPlot.mock.calls[0][0]
    expect(plot.style).toEqual({ flex: 1, height: '100%' })
    expect(plot.config).toEqual({
      toImageButtonOptions: {
        filename: 'TargetEnergy_r_instance_20220927_173640',
      },
      responsive: true,
    })
    expect(plot.layout).toEqual({
      showlegend: true,
      updatemenus: [
        {
          buttons: [
            {
              args: [
                {},
                {
                  xaxis: {
                    title: 'sampling_time[ms]',
                    type: 'log',
                  },
                },
              ],
              label: 'Log Scale',
              method: 'update',
            },
            {
              args: [
                {},
                {
                  xaxis: {
                    title: 'sampling_time[ms]',
                    type: 'linear',
                  },
                },
              ],
              label: 'Linear Scale',
              method: 'update',
            },
          ],
          xanchor: 'left',
          y: 1.2,
          yanchor: 'top',
        },
        {
          buttons: [
            {
              args: [
                {
                  visible: [],
                },
              ],
              label: 'Scatter',
              method: 'update',
            },
            {
              args: [
                {
                  visible: [],
                },
              ],
              label: 'Box',
              method: 'update',
            },
          ],
          x: 0.2,
          xanchor: 'left',
          y: 1.2,
          yanchor: 'top',
        },
      ],
      xaxis: {
        autorange: true,
        title: 'sampling_time[ms]',
        type: 'log',
      },
      yaxis: {
        autorange: true,
        title: 'target_energy',
        type: 'linear',
      },
    })
    expect(plot.data).toHaveLength(1)
  })
})
