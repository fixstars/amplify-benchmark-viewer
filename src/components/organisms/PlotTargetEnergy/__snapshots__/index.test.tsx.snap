// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlotTargetEnergy /> Rendered well 1`] = `
<div>
  <div />
</div>
`;
